vision_evaluation/retrieval_evaluators.py

```diff
@@ -0,0 +1,159 @@
+        """
+        assert average == 'samples'
+        if targets.shape[1] == 1:
+            targets = np.append(targets, np.zeros((targets.shape[0], 1)), axis=1)
+            predictions = np.append(predictions, np.zeros((predictions.shape[0], 1)), axis=1)
+        return sm.precision_score(targets, predictions, average=average)
+
+    def get_report(self):
+        return {self._get_id(): self.calculate_score(average='samples', filter_out_zero_tgt=False)}
+
+
+class RecallAtKEvaluator(RetrievalEvaluator):
+    """
+    Recall evaluator for both multi-class and multi-label classification
+    """
+
+    def __init__(self, k):
+        super(RecallAtKEvaluator, self).__init__(TopKPredictionFilter(k))
+        self.k = k
+
+    def _get_id(self):
+        return f'recall_at_{self.k}'
+
+    def _calculate(self, targets, predictions, average='samples'):
+        """
+        There is a special case that needs to be handled appropriately.
+        Due to filtering conditions, there are occasions where only 1 class remains in targets/predictions and sklearn interpretes this as an invalid configuration for multilabel.
+        Since when average == 'samples' is only calculated for multilabel, sm.recall_score throws an exception. e.g.:
+        predictions = np.array([[True]])
+        targets = np.array([[True]])
+        recall_score(targets, predictions, average='samples') throws the following error:
+        ValueError: Samplewise metrics are not available outside of multilabel classification.
+        """
+        assert average == 'samples'
+        if targets.shape[1] == 1:
+            targets = np.append(targets, np.zeros((targets.shape[0], 1)), axis=1)
+            predictions = np.append(predictions, np.zeros((predictions.shape[0], 1)), axis=1)
+        return sm.recall_score(targets, predictions, average=average)
+
+    def get_report(self):
+        return {self._get_id(): self.calculate_score(average='samples')}
+
+
+class MeanAveragePrecisionAtK(RetrievalEvaluator):
+    """
+    MeanAveragePrecision @ K as defined here:
+    https://stackoverflow.com/questions/54966320/mapk-computation
+    https://medium.com/@misty.mok/how-mean-average-precision-at-k-map-k-can-be-more-useful-than-other-evaluation-metrics-6881e0ee21a9
+    MAP@K is the average of AveP(K) over all queries. Hence it uses average="samples".
+    AveP(K) =  Num / Den, where:
+    Num = sum_i^k P(i) * rel(i), where:
+    P(i) is the Precision @ i
+    rel(i) is an indicator function where rel(i) = 1 if position i is relevant, and rel(i) = 0 otherwise.
+    Den = min(K, number of relevant images)
+    """
+
+    def __init__(self, k):
+        super(MeanAveragePrecisionAtK, self).__init__()
+        self.k = k
+
+    def _get_id(self):
+        return f'map_at_{self.k}'
+
+    def calculate_score(self):
+        if self.k == 0:
+            return 0.0
+        return self._calculate()
+
+    def _calculate(self):
```

Comment:
Can we make sure the arg list conforms with the base class, otherwise if there is change in base class that calls this method, it will fail.

Reply:
@pj-ms I'm not sure I get it. _calculate() is called from calculate_score() defined in the same class so even if _calculate() in base class changes, it will not affect this class. what am I missing? I also think changing def _calculate(self) to def _calculate(self, targets, predictions, average) and then not make use of those args is more confusing.
let me know if you have a different viewpoint

Reply:
For anyone who works with changes in base class, they have to check all implemntations of sub-classes like this one now.

Reply:
A call to _calculate(targets, predictions, average) in MemorizingEverythingEvaluator will fail this class, even thought _calculate was originaly defined in MemorizingEverythingEvaluator, with targets, predictions, average as arguments.